From e26da9937dfb2b3e2adf38b5c2c9521638656d7a Mon Sep 17 00:00:00 2001
From: Jurijs Soloveckis <jsoloveckis@maxlinear.com>
Date: Fri, 4 Aug 2023 11:50:49 +0200
Subject: [PATCH] WLANRTSYS-57531: Improvement for max_listen_interval

Issue: IoT clients disconnects due to inactivity timeout.

Fix:
    1. Our AP only supports listen interval upto 255.
    Hence Set Max listen interval for AP as 255.
    Note: This Will reject assoc for STA with
    listen interval greater than 255.
    2. In inactivity logic, STA disassoc delay to set STA listen interval
    to have time for the STA to sleeping STA to ack the NDP packet.
---
 src/ap/ap_config.c | 2 +-
 src/ap/sta_info.c  | 5 ++++-
 src/ap/sta_info.h  | 1 +
 3 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/ap/ap_config.c b/src/ap/ap_config.c
index 9f95c206f..af9326295 100644
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -101,7 +101,7 @@ void hostapd_config_defaults_bss(struct hostapd_bss_config *bss)
 	bss->ap_protected_keep_alive_required = 0;
 	bss->eapol_version = EAPOL_VERSION;
 
-	bss->max_listen_interval = 65535;
+	bss->max_listen_interval = AP_MAX_LISTEN_INTERVAL;
 
 	bss->pwd_group = 19; /* ECC: GF(p=256) */
 
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index 067e1a982..bae3dcf6a 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -500,6 +500,7 @@ void ap_handle_timer(void *eloop_ctx, void *timeout_ctx)
 	struct sta_info *sta = timeout_ctx;
 	unsigned long next_time = 0;
 	u16 reason;
+	unsigned int disassoc_delay;
 
 	wpa_printf(MSG_DEBUG, "%s: %s: " MACSTR " flags=0x%x timeout_next=%d",
 		   hapd->conf->iface, __func__, MAC2STR(sta->addr), sta->flags,
@@ -624,7 +625,9 @@ skip_poll:
 		wpa_printf(MSG_DEBUG, "%s: register ap_handle_timer timeout "
 			   "for " MACSTR " (%d seconds - AP_DISASSOC_DELAY)",
 			   __func__, MAC2STR(sta->addr), AP_DISASSOC_DELAY);
-		eloop_register_timeout(AP_DISASSOC_DELAY, 0, ap_handle_timer,
+		/* disassoc delay is maximum of listen interval, 3 seconds */
+		disassoc_delay = MAX(AP_DISASSOC_DELAY, (((sta->listen_interval * hostapd_get_beacon_int(hapd)) / 1000) + 1));
+		eloop_register_timeout(disassoc_delay, 0, ap_handle_timer,
 				       hapd, sta);
 		break;
 	case STA_DISASSOC:
diff --git a/src/ap/sta_info.h b/src/ap/sta_info.h
index e9a4c2bb2..dab35f392 100644
--- a/src/ap/sta_info.h
+++ b/src/ap/sta_info.h
@@ -380,6 +380,7 @@ struct sta_info {
 #define AP_MAX_INACTIVITY (5 * 60)
 #define AP_DISASSOC_DELAY (3)
 #define AP_DEAUTH_DELAY (1)
+#define AP_MAX_LISTEN_INTERVAL (255)
 /* Number of seconds to keep STA entry with Authenticated flag after it has
  * been disassociated. */
 #define AP_MAX_INACTIVITY_AFTER_DISASSOC (1 * 30)
-- 
2.43.0

